examples/source_separation/train.py

```diff
@@ -0,0 +1,169 @@
+            "This value is passed as ``init_method`` argument of "
+            "``torch.distributed.init_process_group`` function."
+            'If you are using ``"env://"``, you can additionally configure '
+            'environment variables "MASTER_ADDR" and "MASTER_PORT". '
+            'If you are using ``"file://..."``, then the process has to have '
+            "the access to the designated file. "
+            "See the documentation for ``torch.distributed`` for the detail. "
+            'If you are running the training in a single node, ``"env://"`` '
+            "should do. If you are running the training in multiple nodes, "
+            "you need to provide the file location where all the nodes have "
+            'access, using ``"file://..."`` protocol.'
+        ),
+    )
+    parser.add_argument(
+        "rest", nargs=argparse.REMAINDER, help="Model-specific arguments."
+    )
+    namespace = parser.parse_args(args)
+    if namespace.worker_id is None:
+        if namespace.device_id is not None:
+            raise ValueError(
+                "`--device-id` cannot be provided when runing as master process."
+            )
+        if namespace.num_workers > max_world_size:
+            raise ValueError(
+                "--num-workers ({num_workers}) cannot exceed {device_count}."
+            )
+    if namespace.rest[:1] == ["--"]:
+        namespace.rest = namespace.rest[1:]
+    return namespace
+
+
+def _main(cli_args):
+    args = _parse_args(cli_args)
+
+    if any(arg in ["--help", "-h"] for arg in args.rest):
+        _run_training(args.rest)
+
+    _init_logger(args.worker_id, args.debug)
+    if args.worker_id is None:
+        _run_training_subprocesses(args.num_workers, cli_args)
+    else:
+        dist_utils.setup_distributed(
+            world_size=args.num_workers,
+            rank=args.worker_id,
+            local_rank=args.device_id,
+            init_method=args.sync_protocol,
+        )
+        if torch.cuda.is_available():
+            torch.cuda.set_device(args.device_id)
+            _LG.info("CUDA device set to %s", args.device_id)
+        _run_training(args.rest)
+
+
+def _run_training_subprocesses(num_workers, original_args):
+    workers = []
+    _LG.info("Spawning %s workers", num_workers)
+    for i in range(num_workers):
+        worker_arg = ["--worker-id", f"{i}", "--num-workers", f"{num_workers}"]
+        device_arg = ["--device-id", f"{i}"] if torch.cuda.is_available() else []
+        command = (
+            [sys.executable, "-u", sys.argv[0]]
+            + worker_arg
+            + device_arg
+            + original_args
+        )
+        _LG.info("Launching worker %s: `%s`", i, " ".join(command))
+        worker = subprocess.Popen(command)
+        workers.append(worker)
+
+    num_failed = 0
+    for worker in workers:
+        worker.wait()
+        if worker.returncode != 0:
+            num_failed += 1
+    sys.exit(num_failed)
```

Comment:
I believe this could be accomplished with `torch.multiprocessing.spawn`? Here is [one example on how to do it](https://github.com/pytorch/examples/blob/4b119d735b802453479d739bf823f3f7d8d5d422/imagenet/main.py#L103-L112)

Reply:
Yes, that is right. Adopting `torch.multiprocessing.spawn` makes it harder to understand the code so I stripped it.

Reply:
Also, this entry point is intended only for single-node training mode. (for users who wants to JUST try)

For the real distributed training, please refer to the [README](https://github.com/mthrok/audio/tree/conv-tasnet/examples/source_separation).